File: packages/apl/python/src/apl_api.py
# apl_api.py
"""
Minimal global registries for provider and tool callbacks.
No third-party dependencies; thread-safe via the stdlib `threading` module.
"""

from typing import Callable, Dict, List, Optional
from threading import RLock

ProviderFn = Callable[[str, str, dict, list], str]   # (prompt, model, params, tools) → str
ToolFn     = Callable[..., str]

# ---------- global registries ----------
_providers: Dict[str, ProviderFn] = {}
_tools: Dict[str, ToolFn] = {}

_lock = RLock()


# ---------- provider registry ----------
def add_provider(model: str, fn: ProviderFn) -> None:
    """Register / overwrite a provider callback for a model name."""
    with _lock:
        _providers[model] = fn


def remove_provider(model: str) -> None:
    with _lock:
        _providers.pop(model, None)


def get_provider(model: str) -> Optional[ProviderFn]:
    return _providers.get(model)


def list_providers() -> List[str]:
    return list(_providers.keys())


# ---------- tool registry --------------
def add_tool(name: str, fn: ToolFn) -> None:
    with _lock:
        _tools[name] = fn


def remove_tool(name: str) -> None:
    with _lock:
        _tools.pop(name, None)


def get_tool(name: str) -> Optional[ToolFn]:
    return _tools.get(name)


def list_tools() -> List[str]:
    return list(_tools.keys())

File: packages/apl/typescript/src/apl_api.ts
// apl_api.ts
/**
 * Minimal global registries for provider and tool callbacks.
 * Pure TypeScript, no external packages.
 */

export type ProviderFn = (
  prompt: string,
  model: string,
  params: Record<string, unknown>,
  tools: unknown[],
) => Promise<string> | string;

export type ToolFn = (...args: unknown[]) => Promise<string> | string;

// ---------- global registries ----------
const providers: Record<string, ProviderFn> = {};
const tools: Record<string, ToolFn> = {};

// ---------- provider registry ----------
export function addProvider(model: string, fn: ProviderFn): void {
  providers[model] = fn;
}

export function removeProvider(model: string): void {
  delete providers[model];
}

export function getProvider(model: string): ProviderFn | undefined {
  return providers[model];
}

export function listProviders(): string[] {
  return Object.keys(providers);
}

// ---------- tool registry --------------
export function addTool(name: string, fn: ToolFn): void {
  tools[name] = fn;
}

export function removeTool(name: string): void {
  delete tools[name];
}

export function getTool(name: string): ToolFn | undefined {
  return tools[name];
}

export function listTools(): string[] {
  return Object.keys(tools);
}
